### ISSUE TITLE
Results of umap-learn and UMAP by cuML are different

### ISSUE THREAD

**Issue by JYWangGeneCosmo** [lmcinnes/umap#775]
Hello, 

I was comparing the results produced by your version of umap (**umap-learn**) with the results by **cuML** version of UMAP. I used the same (major) parameters for instantiating UMAP models with two different packages (except some parameters such as `dens_lambda`, `dens_frac`, `dens_var_shift` etc. that are not used in the cuML implementation). When I feed them with the same 100 embeddings, the reduced embeddings have quite different values (see screenshot below). May I ask if you have compared your implementation with **cuML** implementation? It might be hard say, but I hope to know which one is the better implementation (without considering the speed, as cuML leverages GPU), and what could be the major reason that causes this difference (the parameters not used in cuML?).

![image](https://user-images.githubusercontent.com/52869776/135837717-f02fe5e1-4283-443f-a82c-7947df3fbd2b.png)

And It would be really great if you could provide GPU support. May I ask if you have any specific schedule for this? (It's quite difficult to setup cuML-UMAP outside of Google Colab) Thanks a lot! 

Best regards, 
Jingying

**Comment by lmcinnes** [comment]
UMAP as an algorithm is stochastic. As an optimization process it also does not differentiate between rotations or reflections of the same result. Running on many threads changes the stochasticity -- which results get updated in the optimization when, which has flow on effects throughout the optimization. Since cuML UMAP is coded for GPUs and makes use of thousands of threads the final result will not be (and really, cannot be) identical. The results should, however, be qualitiatively very similar.

**Comment by hectorpatino** [comment]
Hi, im having a similar issue and its kind of driving me crazy. I have two notebooks, one of them with scikit-learn api and the other with cuML, both on kaggle. The results of both of them are extremely different.

scikit-learn:  https://www.kaggle.com/code/hectorpatio/umap-parameters/notebook
cuml: https://www.kaggle.com/code/hectorpatio/umap-cuml/notebook

perhaps the most important plot are these ones:

![image](https://user-images.githubusercontent.com/64384750/214158684-22d8e76f-8fb6-49e7-a19d-c1bb2acebff9.png)
scikit-learn

![image](https://user-images.githubusercontent.com/64384750/214158752-3f67bf6a-51c7-445c-8772-a52a83f781b5.png)
cuML.

As you can see they are very different. This is the first time I work with cuML and UMAP, so I don´t really know if its something I did wrong. Any insight about this will be very helpful.

**Comment by beckernick** [comment]
@hectorpatino have you confirmed that the data going into the UMAP calls is the same? The notebooks are slightly different.

I recommend filing a [cuML issue]9https://github.com/rapidsai/cuml/issues/) that includes a minimal, reproducible example separate from Kaggle (if possible)

**Comment by hectorpatino** [comment]
Yeah you are right I updated the notebook last night because I was in a hurry, the version I was talking about was version 2, sorry about that. I will try to use Google to create de reproducible example then. Thank you

**Comment by maciejskorski** [comment]
It's not a data mistake but some issues around scaling and treating outliers.

Under the recent (May'23) Kaggle Image, which provides us with 

```python
from sklearn.datasets import fetch_openml
import matplotlib.pyplot as plt
%matplotlib inline
import umap
import cuml

X,y = fetch_openml('mnist_784', version=1, return_X_y=True, as_frame=False)
y = y.astype(int)

umap_model_1 = umap.UMAP(random_state=42).fit(X)
umap_model_2 = cuml.UMAP(random_state=42).fit(X)

embeds_1 = umap_model_1.transform(X)
embeds_2 = umap_model_2.transform(X)

fig,axs = plt.subplots(1,2)
axs[0].scatter(embeds_1[:,0], embeds_1[:,1], c=y, s=0.1, cmap='Spectral')
axs[1].scatter(embeds_2[:,0], embeds_2[:,1], c=y, s=0.1, cmap='Spectral')
plt.show()
```
I am getting this:
![image](https://github.com/lmcinnes/umap/assets/31315784/d51aaa3d-2971-4346-8727-1e84471cc02b)

Clustering after `cuml` perform worse (statistically) on this example, I have also seen a dataset when this costs as much as ~ 8% in terms of the coherence score. I wonder if this gap can be bridged by controlling more custom options in `cuml`.

**Comment by maciejskorski** [comment]
> @hectorpatino have you confirmed that the data going into the UMAP calls is the same? The notebooks are slightly different.
> 
> I recommend filing a [cuML issue](https://github.com/rapidsai/cuml/issues/) that includes a minimal, reproducible example separate from Kaggle (if possible)

I have provided a good example in my comment (June 9)

**Comment by maciejskorski** [comment]
FYI, I improved the MnWE (minimal not-working example) and shared with the cuML team under [this issue](https://github.com/rapidsai/cuml/issues/5707)
